code/Form/UserFormsRequiredFields.php

```diff
@@ -0,0 +1,173 @@
+<?php
+
+namespace SilverStripe\UserForms\Form;
+
+use SilverStripe\Dev\Debug;
+use SilverStripe\Forms\FileField;
+use SilverStripe\Forms\FormField;
+use SilverStripe\Forms\RequiredFields;
+use SilverStripe\ORM\ArrayLib;
+use SilverStripe\UserForms\Model\EditableFormField;
+
+/**
+ * An extension of RequiredFields which handles conditionally required fields.
+ *
+ * A conditionally required is a field that is required, but can be hidden by display rules.
+ * When it is visible, (according to the submitted form data) it will be validated as required.
+ * When it is hidden, it will skip required validation.
+ *
+ * Required fields will be validated as usual.
+ * Conditionally required fields will be validated IF the display rules are satisfied in the submitted dataset.
+ */
+class UserFormsRequiredFields extends RequiredFields
+{
+    /**
+     * Allows validation of fields via specification of a php function for
+     * validation which is executed after the form is submitted.
+     *
+     * @param array $data
+     *
+     * @return boolean
+     */
+    public function php($data)
```

Comment:
This method is quite long and contains quite some nesting. Abstracting in to separate methods might improve readability